utils/visualize.py

```diff
@@ -0,0 +1,141 @@
+from distutils.util import convert_path
+import networkx as nx
+import matplotlib.pyplot as plt
+from torch_geometric.data import Data, HeteroData
+from torch_geometric.utils import to_networkx
+from torch import Tensor
+import torch
+
+from typing import Union, Optional
+
+
+val_map = {"customer": 1.0, "article": 0.35}
+
+
+def select_layout(G: Union[nx.DiGraph, nx.Graph], node_types: Optional[list]):
+    one_category_nodes = [
+        node if node_type == "customer" else None
+        for node, node_type in zip(G.nodes(), node_types)
+    ]
+
+    pos = nx.bipartite_layout(G, nodes=one_category_nodes)
+    # pos = nx.kamada_kawai_layout(G)
+    # pos = nx.planar_layout(G)
+    # pos = nx.shell_layout(G)
+    # pos = nx.spectral_layout(G)
+    # pos = nx.circular_layout(G)
+    # pos = nx.spiral_layout(G)
+    # pos = nx.spring_layout(G)
+    return pos
+
+
+def get_edges(
+    data: HeteroData,
+) -> tuple[list[tuple[int, int]], list[tuple[int, int]], list[tuple[int, int]]]:
+    _, edge_types = data.metadata()
+
+    edge_dicts = [data[edge_type] for edge_type in edge_types]
+
+    all_edges = edge_dicts[0].edge_index.detach().clone()
+    edge_label = edge_dicts[0].edge_label.detach().clone()
+    edge_label_index = edge_dicts[0].edge_label_index.detach().clone()
+
+    len_customer_indexes = torch.max(all_edges, dim=1)[0][0].item() + 1
+    edge_label_index[1] = torch.add(edge_label_index[1], len_customer_indexes)
```

Comment:
by the way, you can use `+` here, torch has support for overloaded operators on Tensors, just FYI:)

Reply:
no need to change it of course